---
title: "Analysis for Supplementary Information 6"
author: "Alex Huebner"
date: "`21/07/2020"
output:
    github_document:
        fig_width: 8
        fig_height: 6
        toc: true
---

The following script reproduces the analysis used to generate Extended Data Figure 9. It uses the
results generated by the simulation workflow and the general statistics of the contig length, GC
content, mean read length, coverage, and the amount of ancient DNA damage measured as the frequency
of C-to-T substitutions at the 5' read ends.

```{r libraries, echo=F}
library(knitr)
library(data.table)
library(tidyverse)
library(santoku)
library(patchwork)
opts_chunk$set(echo=F, warning=F, message=F, dpi=150, fig.retina = 2)
theme_set(theme_classic(base_size = 8))
```

```{r load_data}
# Statistics of bins assembled from paleofeces in this study
bin_stats <- fread("../results/empiricalbins_summarystats.tsv") %>%
             as_tibble()
# Sequencing length distribution
rl_dist_fns <- list.files("../results/",
                          pattern="_sizefreq\\.size\\.gz",
                          full.names = T)
rl_dist <- map_dfr(rl_dist_fns, function(fn) {
             fread(fn, col.names = c("length", "freq")) %>%
             mutate(readlength = str_replace(str_replace(basename(fn), "_sizefreq\\.size\\.gz", ""),
                                             "rldistribution_", ""))
           }) %>%
           mutate(readlength = factor(readlength, levels = c("short", "medium", "long")))
# GC content
gccontent_genomes <- fread("../results/gccontent.txt")
# Genome length
genomelength <- fread("../results/genomelength.txt")
# MEGAHIT analysis
aln_stats <- fread("../results/alignmentstats_blastn.csv") %>%
             left_join(genomelength, by = c("genome" = "sample"))
mismatches <- fread("../results/mismatches_blastn.csv")
mismatchrate <- fread("../results/mismatchfrac_blastn.csv")
```

# Results

## Panel a: Summary of the assembly statistics of Wibowo et al.

First, we summarised the assembly statistics of `r length(unique(bin_stats$binname))` medium or high
quality bins assembled using MEGAHIT regarding the parameters contig length, GC content, coverage,
and observed C-to-T substitutions on the terminal base at 5' end of the sequencing data. We grouped
the observed values into bins in order to be able to simulate the comparative data more accurately. 

```{r contiglength}
contiglength_stats <- bin_stats %>%
                      mutate(contiglength = contiglength / 1e3,
                             contiglength_bins = chop(contiglength,
                                                      c(0.5, 1, 2, 5, 10, 20, 50, 100, 200, 500),
                                                      drop = F)) %>%
                      group_by(contiglength_bins, .drop = F) %>%
                      count()
contiglength_plt <- ggplot(contiglength_stats,
                           aes(x = contiglength_bins, y = n)) +
                    geom_col(fill = "#40739e") +
                    labs(x = "contig lengths [kb]",
                         y = "number of contigs") +
                    scale_y_continuous(limits = c(0, 45000)) +
                    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
                          axis.title = element_text(size = 7))
```

```{r readlength}
readlength_stats <- bin_stats %>%
                    mutate(RL_bins = chop(meanRL,
                                          seq(0, 300, 25),
                                          drop = F)) %>%
                    group_by(RL_bins, .drop = F) %>%
                    count()
readlength_plt <- ggplot(readlength_stats,
                         aes(x = RL_bins, y = n)) +
                  geom_col(fill = "#40739e") +
                  labs(x = "mean read length [bp]",
                       y = "number of contigs") +
                  scale_y_continuous(limits = c(0, 45000)) +
                  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5, size = 6),
                        axis.title.x = element_text(size = 7),
                        axis.title.y = element_blank())
```


```{r coverage}
coverage_stats <- bin_stats %>%
                  filter(!is.na(coverage)) %>%
                  mutate(coverage_bins = chop(coverage,
                                              c(1, 2, 3, 5, 10, 20, 50, 100, 200, 500, 6000),
                                              drop = F)) %>%
                  group_by(coverage_bins, .drop = F) %>%
                  count()
coverage_plt <- ggplot(coverage_stats,
                       aes(x = coverage_bins, y = n)) +
                geom_col(fill = "#6ab04c") +
                labs(x = "coverage [fold]",
                     y = "number of contigs") +
                scale_y_continuous(limits = c(0, 45000)) +
                theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
                      axis.title.x = element_text(size = 7),
                      axis.title.y = element_blank())
```

```{r damage}
damage_stats <- bin_stats %>%
                filter(!is.na(`CtoT-0`)) %>%
                mutate(damage_bins = chop(`CtoT-0` * 100,
                                          c(0, 1, 2, 2.5, 3, 5, 10, 15, 20, 50, 100),
                                          drop = F)) %>%
                group_by(damage_bins, .drop = F) %>%
                count()
damage_plt <- ggplot(damage_stats,
                     aes(x = damage_bins, y = n)) +
              geom_col(fill = "#c23616") +
              labs(x = "frequency CtoT substitution\nat terminal 5' base [%]",
                   y = "number of contigs") +
              scale_y_continuous(limits = c(0, 45000)) +
              theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
                    axis.title.x = element_text(size = 6),
                    axis.title.y = element_text(size = 7))
```

```{r gccontent}
gccontent_stats <- bin_stats %>%
                   mutate(gc_bins = chop(GC * 100,
                                         c(0, seq(20, 80, 10), 100),
                                         drop = F)) %>%
                   group_by(gc_bins, .drop = F) %>%
                   count()
gccontent_plt <- ggplot(gccontent_stats,
                        aes(x = gc_bins, y = n)) +
                 geom_col(fill = "#e1b12c") +
                 labs(x = "GC content [%]",
                      y = "number of contigs") +
                 scale_y_continuous(limits = c(0, 45000)) +
                 theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5, size = 6),
                       axis.title = element_text(size = 7))
```

```{r wibowo_sumstats_plt, fig.height = 3.5, fig.width = 3.5}
wibowo_sumstats_plt <- gccontent_plt + coverage_plt +
                       damage_plt + readlength_plt +
                       plot_layout(nrow = 2, ncol = 2)
wibowo_sumstats_plt
```

**Extended Data Figure 9a**: **Overview of the number of assembled contigs per bin with respect to
the contig length, the GC content, the coverage along the contig, and the CtoT substitution
frequency on the terminal base at the 5' end of the sequencing data.** `r nrow(bin_stats)` from
`r length(unique(bin_stats$binname))` medium or high quality bins were summarised.

## Panel b - Overview of the simulation scheme

Based on the observed values in bins assembled in this study, we simulated short-read sequencing
data with ancient DNA damage using gargammel (Renauld et al., 2017).

```{r gccontent_genomes}
gccontent_genomes_plt <- gccontent_genomes %>%
                         rename(genome = sample) %>%
                         filter(genome %in% c("Msmithii", "Tforsythia", "Adentalis")) %>%
                         mutate(genome = str_replace(genome, "^([A-Z])", "\\1. ")) %>%
                         mutate(genome = factor(genome, levels = .$genome),
                                label = str_c(round(GC * 100, 1), " %")) %>%
                         ggplot(aes(x = genome, y = GC)) +
                         geom_col(fill = "#e1b12c") +
                         geom_text(aes(y = GC + 0.1, label = label), size = 1.4) +
                         labs(x = "genome",
                              y = "GC content") +
                         scale_y_continuous(limits = c(0, 1), labels = scales::percent_format(),
                                            breaks = c(0.25, 0.50, 0.75, 1)) +
                         theme(axis.text.x = element_text(angle = 90, hjust = .5, vjust = 0.5),
                               axis.title = element_text(size = 6),
                               axis.text.y = element_blank(),
                               axis.ticks.y = element_blank())
```

```{r readlengthdist}
readlengthdist_plt <- ggplot(rl_dist,
                             aes(x = length, y = freq)) +
                      geom_col(fill = "#40739e", colour = "#40739e") +
                      facet_wrap(~ readlength, nrow = 3) +
                      labs(x = "read length [bp]",
                           y = "simulated read length distribution") +
                      theme(axis.title = element_text(size = 6),
                            strip.text = element_text(size = 6),
                            axis.text.y = element_text(size = 5))
```

```{r simulated_coverage}
simulated_coverage <- tibble(label = c("[5, 10)", "[10, 20)", "[20, 50)", "[50, 100)", "[100, 100)"),
                             start = c(5, 10, 20, 50, 98),
                             end = c(10, 20, 50, 100, 102)) %>%
                      mutate(label = factor(label, levels = rev(.$label)))

simulated_coverage_plt <- ggplot(simulated_coverage,
                                 aes(x = label)) +
                          geom_linerange(aes(ymin = start, ymax = end),
                                         size = 2.3, colour = "#6ab04c") +
                          geom_text(aes(y = end + 7.5, label = label), hjust = 0, size = 1.5) +
                          coord_flip() +
                          labs(x = "simul. coverage bins",
                               y = "coverage [fold]") +
                          scale_y_continuous(limits = c(0, 150)) +
                          theme(axis.text.y = element_blank(),
                                axis.ticks.y = element_blank(),
                                axis.title = element_text(size = 6))
```

```{r simulated_damage}
damage_levels <- tibble(label = c("0%", "1%", "1.5%", "2%", "2.5%", "3%", "5%", "10%", "15%", "20%"),
                        damage = c(0, 0.01, 0.015, 0.02, 0.025, 0.03, 0.05, 0.1, 0.15, 0.2)) %>%
                 mutate(label = factor(label, levels = rev(.$label)))

simulated_damage_plt <- ggplot(damage_levels,
                               aes(x = label)) +
                        geom_linerange(aes(ymax = damage), ymin = 0,
                                       size = 1.8, colour = "#c23616") +
                        geom_text(aes(y = damage + 0.0035, label = label), hjust = 0, size = 1.4) +
                        coord_flip() +
                        labs(x = "simul. aDNA damage bins",
                             y = "frequency of CtoT substitution\nat terminal 5' base [%]") +
                        scale_y_continuous(limits = c(-0.01, 0.22),
                                           breaks = c(0, 0.1, 0.2),
                                           labels = scales::percent_format()) +
                        theme(axis.text.y = element_blank(),
                              axis.ticks.y = element_blank(),
                              axis.title = element_text(size = 6))
```

```{r simulated_contiglength}
simulated_contiglength <- tibble(label = c("[0.5, 1)", "[1, 2)", "[2, 5)",
                                           "[5, 10)", "[10, 20)", "[20, 50)",
                                           "[50, 100)", "[100, 200)", "[200, 500)"),
                                 start = c(0.5, 1, 2, 5, 10, 20, 50, 100, 200),
                                 end = c(1, 2, 5, 10, 20, 50, 100, 200, 500)) %>%
                          mutate(label = factor(label, levels = rev(.$label)))

simulated_contiglength_plt <- ggplot(simulated_contiglength,
                                     aes(x = label)) +
                              geom_linerange(aes(ymin = start, ymax = end),
                                             size = 2.5, colour = "#40739e") +
                              geom_text(aes(y = end + 2.5, label = label), hjust = 0, size = 1.4) +
                              coord_flip() +
                              labs(x = "simulated contig length bins",
                                   y = "contig length [kb]") +
                              scale_y_continuous(limits = c(0, 650)) +
                              theme(axis.text.y = element_blank(),
                                    axis.ticks.y = element_blank(),
                                    axis.title = element_text(size = 6))
```

```{r simulation_scheme_plt, fig.height = 3.5, fig.width = 3.5}
simulation_scheme_plt <- ((gccontent_genomes_plt /
                           simulated_coverage_plt /
                           simulated_damage_plt) |
                           readlengthdist_plt) 
simulation_scheme_plt
```

**Extended Data Figure 9b**: **Overview of the parameter space of the variables GC content,
sequencing depth, observed aDNA damage, and read length that was used for simulating short-read
sequencing using gargammel.**


## Panel c - Mismatch rate

```{r mismatchrate_plt, fig.height = 3, fig.width = 7}
# Simulation parameters
covbins <- tibble(coverage = c("[0, 1)", "[1, 2)", "[2, 5)", "[5, 10)",
                            "[10, 20)", "[20, 50)", "[50, 100)", "[100, 100)"),
                  covbin = seq(0, 7))
damagelevels <- tibble(damage = c("0", "1", "1.5", "2", "2.5", "3", "5", "10", "15", "20"),
                       damagelevel = seq(0, 9))

mismatchrate_summary <- mismatchrate %>%
                        filter(covbin > 2) %>%
                        left_join(covbins, by = "covbin") %>%
                        left_join(damagelevels, by = "damagelevel") %>%
                        left_join(genomelength, by = c("genome" = "sample")) %>%
                        mutate(genome = str_replace(genome, "^([A-Z])", "\\1\\. "),
                               genome = factor(genome, levels = c("M. smithii", "T. forsythia", "A. dentalis")),
                               rl = factor(rl, levels = c("short", "medium", "long")),
                               coverage = factor(coverage, levels = covbins$coverage),
                               damage = factor(damage, levels = damagelevels$damage),
                               normRate = rate / contiglength * 1000) %>%
                        group_by(genome, rl, coverage, damage) %>%
                        summarise(meanRate = mean(normRate),
                                  medianRate = median(normRate),
                                  UQRate = quantile(normRate, 0.95))

mismatchrate_plt <- mismatchrate_summary %>%
                    ggplot(aes(x = damage, y = coverage, fill = UQRate)) +
                    geom_tile() +
                    facet_grid(rl ~ genome) +
                    labs(fill = "95% quantile of\nmismatches\nper 1 kb",
                         x = "damage [%]",
                         y = "coverage bin") +
                    scale_fill_gradient(label = scales::number_format(accuracy = 1),
                                        breaks = seq(0, 10, 2),
                                        high = "#132B43", low = "#56B1F7") +
                    theme(legend.position = "right")
mismatchrate_plt
```

**Extended Data Figure 9c**: **Number of mismatches per 1 kb of alignable contig sequence with respect to the
reference genome as observed at the 95% quantile for all combinations of reference genome, read
length distribution, simulated aDNA damage, and coverage averaged across the five replicates.**

## Panel d - log2-ratio between C-to-T and other substitutions

```{r mismatches_logratio_plt, fig.height = 3, fig.width = 7}
mismatches_summary <- mismatches %>%
                      filter(covbin > 2) %>%
                      left_join(covbins, by = "covbin") %>%
                      left_join(damagelevels, by = "damagelevel") %>%
                      mutate(genome = str_replace(genome, "^([A-Z])", "\\1\\. "),
                             genome = factor(genome, levels = c("M. smithii", "T. forsythia", "A. dentalis")),
                             rl = factor(rl, levels = c("short", "medium", "long")),
                             coverage = factor(coverage, levels = covbins$coverage),
                             damage = factor(damage, levels = damagelevels$damage)) %>%
                      pivot_longer(AC:Ins, names_to = "operation", values_to = "count") %>%
                      mutate(op_type = ifelse(operation == "CT", operation,
                                              ifelse(operation %in% c("Del", "Ins"), "Indel", "other_subst"))) %>%
                      group_by(genome, rl, damage, coverage, op_type) %>%
                      summarise(meanCount = mean(count)) %>%
                      left_join(aln_stats %>%
                                select(genome, rl, coverage, damage, alignmentlength) %>%
                                group_by(genome, rl, coverage, damage) %>%
                                summarise(alignmentlength = mean(alignmentlength))) %>%
                      pivot_wider(names_from = "op_type", values_from = "meanCount") %>%
                      mutate(logratio = log2(CT / other_subst))

mismatches_plt <- mismatches_summary %>%
                  ggplot(aes(x = damage, y = coverage, fill = logratio)) +
                  geom_tile() +
                  facet_grid(rl ~ genome) +
                  labs(fill = "log2 ratio\n(CT / other subst.)",
                       x = "damage [%]",
                       y = "coverage bin") +
                  scale_fill_gradient2(low = scales::muted("blue"), high = scales::muted("red"), na.value = "grey80", midpoint = 0,
                                       limits = c(-3, 3)) +
                  theme(legend.position = "right")
mismatches_plt
```

**Extended Data Figure 9d**: **Log2-ratio of C-to-T substitutions to the average number of all other substitutions
per 1 kb of alignable contig sequence for all combinations of reference genome, read length
distribution, simulated aDNA damage, and coverage averaged across the five replicates. Positive
values indicate an excess of C-to-T substitutions.**
